mlserver/batch_processing.py

```diff
@@ -0,0 +1,283 @@
+            index += 1
+
+
+async def consume(
+    model_name: str,
+    worker_id: int,
+    triton_client: httpclient.InferenceServerClient,
+    binary_data: bool,
+    extra_verbose: bool,
+    queue_in: asyncio.Queue,
+    queue_out: asyncio.Queue,
+):
+    while True:
+        input_item = await queue_in.get()
+        try:
+            output_item = await process_item(
+                input_item, model_name, worker_id, triton_client, binary_data
+            )
+        except Exception:
+            if extra_verbose:
+                logger.error(
+                    f"consumer {worker_id}: failed to process item {input_item}"
+                )
+            queue_in.task_done()
+            continue
+
+        await queue_out.put(output_item)
+        queue_in.task_done()
+
+
+async def finalize(queue: asyncio.Queue, fname: str):
+    # TODO: Test if output directory is writtable, sysexit otherwise.
+    async with aiofiles.open(fname, "wb") as f:
+        while True:
+            item = await queue.get()
+            try:
+                output = serialize_triton_infer_result(item)
+                await f.write(output)
+                await f.write(b"\n")
+            except Exception as e:
+                logger.error(f"Failed to finalize task: {e}")
+            queue.task_done()
+
+
+async def process_batch(
+    model_name,
+    url,
+    workers,
+    input_data_path,
+    output_data_path,
+    binary_data,
+    transport,
+    use_ssl,
+    insecure,
+    verbose,
+    extra_verbose,
+):
+    start_time = timer()
+
+    if transport == "grpc":
+        raise click.BadParameter("The 'grpc' transport is not yet supported.")
+
+    if extra_verbose:
+        verbose = True
+        logger.info("Running in extra verbose mode.")
+    elif verbose:
+        logger.info("Running in verbose mode.")
+
+    setup_logging(debug=verbose)
+    logger.info(f"Server url: {url}")
+    logger.info(f"input file path: {input_data_path}")
+    logger.info(f"output file path: {output_data_path}")
+
+    if insecure:
+        ssl_context = False
+    else:
+        ssl_context = None
+
+    triton_client = httpclient.InferenceServerClient(
+        url=url,
+        verbose=extra_verbose,
+        conn_limit=workers,
+        ssl=use_ssl,
+        ssl_context=ssl_context,
+    )
+
+    queue_in = asyncio.Queue(2 * workers)
+    queue_out = asyncio.Queue(2 * workers)
+
+    consumers = []
+    for worker_id in range(workers):
+        consumer = asyncio.create_task(
+            consume(
+                model_name,
+                worker_id,
+                triton_client,
+                binary_data,
+                extra_verbose,
+                queue_in,
+                queue_out,
+            )
+        )
+        consumers.append(consumer)
+
+    finalizer = asyncio.create_task(finalize(queue_out, output_data_path))
+
+    await produce(queue_in, input_data_path)
+    await queue_in.join()
+    await queue_out.join()
```

Comment:
May not be correct, but would there be a case where the out queue finishes and consumers/producers/finalizers get cancelled before the final input gets written to file (eg in the case of a super long response for example)

Reply:
Nevermind, spoke too soon - the queue is only finalised when the task is marked as done which happens after writing file so should not be an issue

Reply:
More info for completeness:
- https://docs.python.org/3/library/asyncio-queue.html#asyncio.Queue.join
- https://docs.python.org/3/library/asyncio-queue.html#examples

Reply:
So there should be nothing to worry about here :+1:. The logic closely follow official queue example from Python documentation.